### ISSUE TITLE
[KLUE] 오피스 아워(0930) 정리

### ISSUE THREAD

**Issue by sangmandu** [sangmandu/SangSangPlus#101]
1. 모델 비교
- KLUE-RoBERTa
- mBERT
- KoBERT
- GPTBERT
- ? 뭐였떠라
- ? 못봤어용
 
2. Entity special token을 추가해보자.
  
![image](https://user-images.githubusercontent.com/45033215/135429623-f9ec9398-797a-49d4-aaae-0b6c9d1d7634.png)

![image](https://user-images.githubusercontent.com/45033215/135429629-e951b468-4931-4f26-8c3a-a616733d0d6c.png)

 
special token 사용하지 말고 문장부호를 이용해보자!! 이게 성능이 제일 높으니까!! 근데 모델마다 다르니까 special token도 사용해보자
또는, entity 유무에 따라 임베딩 레이어를 추가해보자!! 토큰이나 특수문자를 쓰지 않는 방법
 
![image](https://user-images.githubusercontent.com/45033215/135429641-6e867275-09b1-4590-98c4-ed9e9acfd25e.png)

3. 모델과 분류기 사이를 더 깊게 쌓아보자.
 
![image](https://user-images.githubusercontent.com/45033215/135429650-ba74adce-c175-4dc3-94a0-11736a0778fc.png)

 
4. pretrain 모델을 더 학습을 하자. 단순히 fine tuning이 아니라 mask 토큰들을 추가해서 이걸 하자!! 진짜 pretraining task처럼
 
![image](https://user-images.githubusercontent.com/45033215/135429664-96351416-3e43-4f9c-944b-91c3ddfc7f2c.png)

 
5. data augmentation
 
![image](https://user-images.githubusercontent.com/45033215/135429702-59fbd739-efbf-4737-b99a-404790fad450.png)

augmentation 에 관한 논문
EDA : Easy Data Augmentation => https://github.com/toriving/KoEDA
 
![image](https://user-images.githubusercontent.com/45033215/135429718-c973f05b-55d5-4274-ae4d-eae3812c878e.png)

카카오의 pororo 라이브러리와 KoEDA를 사용하기!!!
 
![image](https://user-images.githubusercontent.com/45033215/135429733-0d1dc712-9e15-47a1-802e-f89cbf03998a.png)

작은 데이터셋에서는 상승폭이 더 높음 우리의 프로젝트에서도 잘 작용할 것으로 보임. 전체 데이터셋을 사용하더라도 성능이 오르기 때문에 무조건 해봐야 할 부분

AEDA 
 
![image](https://user-images.githubusercontent.com/45033215/135429743-54226cf4-47dc-48fa-8328-3ba52263551d.png)

이것도 KoEDA에 같이 구현되어있다.
위 그래프를 보면 성능이 AEDA가 더 높음. 아마도 사용한다면 EDA와 AEDA를 같이 사용하기보다는 AEDA만을 사용하는 것이 좋아보임

그리고, 이거를 데이터 불균형을 해결하기 위해 일단 모두 하고, 추가적으로 하면 어떨까? (오히려 안좋아질 수도 있음) 일단 해볼만한 아이디어?
 
![image](https://user-images.githubusercontent.com/45033215/135429764-8b7cc529-5903-4343-9f74-5bc2c6b3c7e5.png)

포로로 라이브러리를 이용해서 이것이 가능하다.
 
![image](https://user-images.githubusercontent.com/45033215/135429773-ef5694dc-0417-474e-b989-d9deeeb33973.png)

한국어-영어-한국어도 되고 한국어-일본어-한국어도 되고 한국어-영어-일본어-한국어로 바꿔도 된다. 포로로 라이브러리가 아니더라도 구글 번역기, 파파고에서 얻은 결과를 가지고 사용할 수도 있다.

 
![image](https://user-images.githubusercontent.com/45033215/135429794-0f5b6d9c-e5bc-47b0-af79-dab477b39304.png)

어떤 언어로 변환해야 잘 되는지 실험 또는 조사해봐야한다!!
 
6. 
 
![image](https://user-images.githubusercontent.com/45033215/135429831-b3280cdb-03e7-42a2-b1af-42d1f045da2e.png)

엔티티가 등장할 때 마다 SEP 추가하기
 
![image](https://user-images.githubusercontent.com/45033215/135429844-df4a5ee6-f484-4566-ac4f-f04a26c025e0.png)

버트는 싱글인 경우는 SEP 한 개, 멀티인 경우는 SEP 두개를 사용한다.
 
![image](https://user-images.githubusercontent.com/45033215/135429859-f852b776-ffb3-4393-b9ba-a9a80d7a020a.png)

FINE TUNING도 PRE TRAIN이랑 최대한 비슷하게하자!! 지금 베이스라인에서 제공하는 방법은 ENTITY의 뒤에 SEP을 추가한다!!

 
7. Q&A
### 1.
![image](https://user-images.githubusercontent.com/45033215/135429872-8e79afd2-5854-47c0-b895-40743f0045b8.png)

 ### 2.
![image](https://user-images.githubusercontent.com/45033215/135429903-f74be1db-5469-446f-b664-f5b7a580afbe.png)

그런거 없음 그냥 공부에 도움이 될까봐 추가한 것임

 ### 3.
![image](https://user-images.githubusercontent.com/45033215/135429912-9b79c15f-5f1b-48ee-adc9-a6a132b00b4f.png)

생성된 것중에 일부는 버려도 괜찮음.
내 아이디어 : 고유명사(해당 entity의 속성)로 대체하고 다시 대체하기
 ### 4.
![image](https://user-images.githubusercontent.com/45033215/135429940-8ee47601-4d16-42aa-82f3-f72979408e95.png)
![image](https://user-images.githubusercontent.com/45033215/135430771-14944b34-f10d-4d20-8da9-407c31d5a899.png)

robust한 모델을 위해서라도(일반화 성능을 이야기하는 듯) 냅두는 것이 좋아보임. 근데 이게 매우 적은 클래스의 데이터라면 바꿔도 좋을 수도 있는데 이 부분은 실험적으로 해봐야 하는 거 같음

 ### 5.
![image](https://user-images.githubusercontent.com/45033215/135430260-a99181f1-9e23-4b40-b0b7-9345a0eda162.png)
klue 데이터셋처럼 동일하게 한 것
 ### 6.
![image](https://user-images.githubusercontent.com/45033215/135430326-24648351-2f62-4595-b1ad-104f6c4128ff.png)
걍 entity 위치 잡으려고 있는 것
 ### 7.
![image](https://user-images.githubusercontent.com/45033215/135430016-f2c17185-a302-4511-a32b-629e232fc020.png)

딱히 그런건 아님. 그냥 라벨링 할 때 라벨링 한 사람이 어느 단어를 보고 라벨링 했느냐에 대한 부분.
### 8.
![image](https://user-images.githubusercontent.com/45033215/135430169-996e8c34-677e-4659-9935-55e6a67e875f.png)

의미가 달라질 수 있으니까 그 부분 유의해서 바꿔봐라. (별로 추천하는 뉘앙스는 아닌듯)
특정 relation의 경우(ex member-member) 괜찮음. 그러나 안되는 경우도 있으니 주의(홍길동은 20살, 20살은 홍길동)
### 9.
![image](https://user-images.githubusercontent.com/45033215/135430720-dd34c4c0-a4c4-4b91-ac98-4f41d05617b2.png)
추후에 klue 채널에 답변 남겨줄게

### 10.
![image](https://user-images.githubusercontent.com/45033215/135430851-cd92fe59-8876-4c90-bbab-0b0e0bd1d024.png)
좋아보임. 
=> 그리고 이게 뭔가 EDA에서 동의어로 치환하는 것 뿐만 아니라 유형이 같은 단어로 바꾸는 거라 Augmentation에도 좋아보임. 근데 좀 잘 바뀐다는 가정하에서.
### 11.
![image](https://user-images.githubusercontent.com/45033215/135431125-61cb7edb-c0da-4901-83c6-bf33ae69f876.png)
- unknown 토큰이 있어야 test시에도 적용하니까 학습시에도 이런게 필요함 ㅇㅇ
- entitiy에 unknown 토큰이 있으면 그건 그거 나름대로의 능력을 기르는 거기 때문에 테스트 시에도 잘 대응할 수 있을거임 (근데 우리는 그런 테스트셋이 없으니까 안해도 될듯)
이에대한 멘토님 답변
![image](https://user-images.githubusercontent.com/45033215/135432940-a080fe7c-d5a6-4dc4-9e9d-6455728bd711.png)

### 추가적인 멘토님 답변
![image](https://user-images.githubusercontent.com/45033215/135431903-f07b25a2-30ee-4ccb-9013-185e6d6f5833.png)

![image](https://user-images.githubusercontent.com/45033215/135431989-4e7ae7b0-3202-4911-b836-fc6bd64fa9db.png)

### 12.
![image](https://user-images.githubusercontent.com/45033215/135432149-73688e12-25f1-406b-b978-29f47245c6e3.png)
모델이 들어갈 때는 토크나이징 이후에 결과가 입력되니까 토크나이징 된 길이를 보는게 좋아보임

### 13.
![image](https://user-images.githubusercontent.com/45033215/135435425-9e09abaa-0240-484a-948d-258173342c78.png)

케바케임

### 14.
![image](https://user-images.githubusercontent.com/45033215/135435278-56a7a7da-01b7-48eb-862a-a5c626d38909.png)

??(잘 못들음)가 제일 좋을 것 같긴한데, 이건 힘드니까 걍 validation으로 downstream? 1~2 에폭만 돌려서 FINETUNING??
![image](https://user-images.githubusercontent.com/45033215/135435343-5d78d92b-becc-42df-a195-2d65b0e72ee8.png)

### 15.
![image](https://user-images.githubusercontent.com/45033215/135435175-5eacd254-2f48-43f6-8c50-fdea11fd6a4e.png)
비추

### 16.
![image](https://user-images.githubusercontent.com/45033215/135435671-5edce09c-ae9f-4877-a5ea-2a7ca068eadb.png)
비추
